### PR TITLE
spread errors on incremental responses

### LINKED ISSUE #1818: @defer: error related to a field in a deferred fragment appears in chunk for non-deferred fragment
**Describe the bug**
Using the graphql [test suite](https://github.com/graphql/graphql-js/blob/a24a9f35b876bdd0d5050eca34d3020bd0db9a29/src/execution/__tests__/defer-test.ts#L424) as an inspiration, I am selecting a field that resolves to an error, in a deferred fragment:

```graphql
query HandlesErrorsThrownInDeferredFragmentsQuery {
  computer(id: "Computer1") {
    id
    ...ComputerErrorField @defer
  }
}

fragment ComputerErrorField on Computer {
  errorField
}
```

I am receiving the error in the first chunk (along with `id`) and no errors in the second chunk, whereas I would have expected the opposite (error in the 2nd chunk along `errorField`), as graphql-js apparently does.

Unsure if this is the spec really mandates this, it may be just a different and acceptable implementation, if so feel free to close!

```
--graphql
content-type: application/json

{"data":{"computer":{"id":"Computer1"}},"errors":[{"message":"Subgraph errors redacted"}],"hasNext":true}
--graphql
content-type: application/json

{"hasNext":false,"incremental":[{"data":{"errorField":null},"path":["computer"]}]}
--graphql--
```

**To Reproduce**

```
git clone git@github.com:apollographql/apollo-kotlin.git
cd apollo-kotlin && git checkout defer-with-router-tests
(cd tests/defer/router/subgraphs/computers && npm install && APOLLO_PORT=4001 npm start)&
path/to/router --supergraph tests/defer/router/simple-supergraph.graphqls &
curl --request POST     --header 'content-type: application/json'     --url http://localhost:4000     --data '{"query":"query HandlesErrorsThrownInDeferredFragmentsQuery {  computer(id: \"Computer1\") {    id    ...ComputerErrorField @defer  } }  fragment ComputerErrorField on Computer {  errorField }"}' --header 'Accept: multipart/mixed; deferSpec=20220824'
```

### LINKED ISSUE #1834: @defer: duplicated errors across incremental items
**Describe the bug**
In https://github.com/apollographql/router-defer-e2e-tests/pull/8, some tests were added for error cases.

For the following query:

```graphql
query TestQuery {
  allProducts {
    ...ErrorFragment @defer
    sku
    id
  }
}
fragment ErrorFragment on Product {
  promiseNonNullErrorField
}
```

the response is

```

--graphql
content-type: application/json

{"data":{"allProducts":[{"__typename":"Product","sku":"federation","id":"apollo-federation"},{"__typename":"Product","sku":"studio","id":"apollo-studio"},{"__typename":"Product","sku":"client","id":"apollo-client"}]},"hasNext":true}
--graphql
content-type: application/json

{"hasNext":false,"incremental":[{"data":null,"path":["allProducts",0],"errors":[{"message":"Cannot return null for non-nullable field Product.promiseNonNullErrorField.","locations":[{"line":1,"column":126}],"path":["allProducts","@","_entities",0,"promiseNonNullErrorField"],"extensions":{"code":"INTERNAL_SERVER_ERROR","exception":{"stacktrace":["Error: Cannot return null for non-nullable field Product.promiseNonNullErrorField.","    at completeValue (/usr/src/app/node_modules/graphql/execution/execute.js:594:13)","    at /usr/src/app/node_modules/graphql/execution/execute.js:486:9","    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)","    at async Promise.all (index 1)","    at async Promise.all (index 0)","    at async Promise.all (index 0)","    at async execute (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:205:20)","    at async processGraphQLRequest (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:148:28)","    at async processHTTPRequest (/usr/src/app/node_modules/apollo-server-core/dist/runHttpQuery.js:220:30)"]}}},{"message":"Cannot return null for non-nullable field Product.promiseNonNullErrorField.","locations":[{"line":1,"column":126}],"path":["allProducts","@","_entities",1,"promiseNonNullErrorField"],"extensions":{"code":"INTERNAL_SERVER_ERROR","exception":{"stacktrace":["Error: Cannot return null for non-nullable field Product.promiseNonNullErrorField.","    at completeValue (/usr/src/app/node_modules/graphql/execution/execute.js:594:13)","    at /usr/src/app/node_modules/graphql/execution/execute.js:486:9","    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)","    at async Promise.all (index 1)","    at async Promise.all (index 1)","    at async Promise.all (index 0)","    at async execute (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:205:20)","    at async processGraphQLRequest (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:148:28)","    at async processHTTPRequest (/usr/src/app/node_modules/apollo-server-core/dist/runHttpQuery.js:220:30)"]}}},{"message":"Cannot return null for non-nullable field Product.promiseNonNullErrorField.","locations":[{"line":1,"column":126}],"path":["allProducts","@","_entities",2,"promiseNonNullErrorField"],"extensions":{"code":"INTERNAL_SERVER_ERROR","exception":{"stacktrace":["Error: Cannot return null for non-nullable field Product.promiseNonNullErrorField.","    at completeValue (/usr/src/app/node_modules/graphql/execution/execute.js:594:13)","    at /usr/src/app/node_modules/graphql/execution/execute.js:486:9","    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)","    at async Promise.all (index 1)","    at async Promise.all (index 2)","    at async Promise.all (index 0)","    at async execute (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:205:20)","    at async processGraphQLRequest (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:148:28)","    at async processHTTPRequest (/usr/src/app/node_modules/apollo-server-core/dist/runHttpQuery.js:220:30)"]}}},{"message":"Cannot return null for non-nullable field ProductItf.promiseNonNullErrorField","path":["allProducts",0]},{"message":"Cannot return null for non-nullable field ProductItf.promiseNonNullErrorField","path":["allProducts",1]},{"message":"Cannot return null for non-nullable field ProductItf.promiseNonNullErrorField","path":["allProducts",2]}]},{"data":null,"path":["allProducts",1],"errors":[{"message":"Cannot return null for non-nullable field Product.promiseNonNullErrorField.","locations":[{"line":1,"column":126}],"path":["allProducts","@","_entities",0,"promiseNonNullErrorField"],"extensions":{"code":"INTERNAL_SERVER_ERROR","exception":{"stacktrace":["Error: Cannot return null for non-nullable field Product.promiseNonNullErrorField.","    at completeValue (/usr/src/app/node_modules/graphql/execution/execute.js:594:13)","    at /usr/src/app/node_modules/graphql/execution/execute.js:486:9","    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)","    at async Promise.all (index 1)","    at async Promise.all (index 0)","    at async Promise.all (index 0)","    at async execute (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:205:20)","    at async processGraphQLRequest (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:148:28)","    at async processHTTPRequest (/usr/src/app/node_modules/apollo-server-core/dist/runHttpQuery.js:220:30)"]}}},{"message":"Cannot return null for non-nullable field Product.promiseNonNullErrorField.","locations":[{"line":1,"column":126}],"path":["allProducts","@","_entities",1,"promiseNonNullErrorField"],"extensions":{"code":"INTERNAL_SERVER_ERROR","exception":{"stacktrace":["Error: Cannot return null for non-nullable field Product.promiseNonNullErrorField.","    at completeValue (/usr/src/app/node_modules/graphql/execution/execute.js:594:13)","    at /usr/src/app/node_modules/graphql/execution/execute.js:486:9","    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)","    at async Promise.all (index 1)","    at async Promise.all (index 1)","    at async Promise.all (index 0)","    at async execute (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:205:20)","    at async processGraphQLRequest (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:148:28)","    at async processHTTPRequest (/usr/src/app/node_modules/apollo-server-core/dist/runHttpQuery.js:220:30)"]}}},{"message":"Cannot return null for non-nullable field Product.promiseNonNullErrorField.","locations":[{"line":1,"column":126}],"path":["allProducts","@","_entities",2,"promiseNonNullErrorField"],"extensions":{"code":"INTERNAL_SERVER_ERROR","exception":{"stacktrace":["Error: Cannot return null for non-nullable field Product.promiseNonNullErrorField.","    at completeValue (/usr/src/app/node_modules/graphql/execution/execute.js:594:13)","    at /usr/src/app/node_modules/graphql/execution/execute.js:486:9","    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)","    at async Promise.all (index 1)","    at async Promise.all (index 2)","    at async Promise.all (index 0)","    at async execute (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:205:20)","    at async processGraphQLRequest (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:148:28)","    at async processHTTPRequest (/usr/src/app/node_modules/apollo-server-core/dist/runHttpQuery.js:220:30)"]}}},{"message":"Cannot return null for non-nullable field ProductItf.promiseNonNullErrorField","path":["allProducts",0]},{"message":"Cannot return null for non-nullable field ProductItf.promiseNonNullErrorField","path":["allProducts",1]},{"message":"Cannot return null for non-nullable field ProductItf.promiseNonNullErrorField","path":["allProducts",2]}]},{"data":null,"path":["allProducts",2],"errors":[{"message":"Cannot return null for non-nullable field Product.promiseNonNullErrorField.","locations":[{"line":1,"column":126}],"path":["allProducts","@","_entities",0,"promiseNonNullErrorField"],"extensions":{"code":"INTERNAL_SERVER_ERROR","exception":{"stacktrace":["Error: Cannot return null for non-nullable field Product.promiseNonNullErrorField.","    at completeValue (/usr/src/app/node_modules/graphql/execution/execute.js:594:13)","    at /usr/src/app/node_modules/graphql/execution/execute.js:486:9","    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)","    at async Promise.all (index 1)","    at async Promise.all (index 0)","    at async Promise.all (index 0)","    at async execute (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:205:20)","    at async processGraphQLRequest (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:148:28)","    at async processHTTPRequest (/usr/src/app/node_modules/apollo-server-core/dist/runHttpQuery.js:220:30)"]}}},{"message":"Cannot return null for non-nullable field Product.promiseNonNullErrorField.","locations":[{"line":1,"column":126}],"path":["allProducts","@","_entities",1,"promiseNonNullErrorField"],"extensions":{"code":"INTERNAL_SERVER_ERROR","exception":{"stacktrace":["Error: Cannot return null for non-nullable field Product.promiseNonNullErrorField.","    at completeValue (/usr/src/app/node_modules/graphql/execution/execute.js:594:13)","    at /usr/src/app/node_modules/graphql/execution/execute.js:486:9","    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)","    at async Promise.all (index 1)","    at async Promise.all (index 1)","    at async Promise.all (index 0)","    at async execute (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:205:20)","    at async processGraphQLRequest (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:148:28)","    at async processHTTPRequest (/usr/src/app/node_modules/apollo-server-core/dist/runHttpQuery.js:220:30)"]}}},{"message":"Cannot return null for non-nullable field Product.promiseNonNullErrorField.","locations":[{"line":1,"column":126}],"path":["allProducts","@","_entities",2,"promiseNonNullErrorField"],"extensions":{"code":"INTERNAL_SERVER_ERROR","exception":{"stacktrace":["Error: Cannot return null for non-nullable field Product.promiseNonNullErrorField.","    at completeValue (/usr/src/app/node_modules/graphql/execution/execute.js:594:13)","    at /usr/src/app/node_modules/graphql/execution/execute.js:486:9","    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)","    at async Promise.all (index 1)","    at async Promise.all (index 2)","    at async Promise.all (index 0)","    at async execute (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:205:20)","    at async processGraphQLRequest (/usr/src/app/node_modules/apollo-server-core/dist/requestPipeline.js:148:28)","    at async processHTTPRequest (/usr/src/app/node_modules/apollo-server-core/dist/runHttpQuery.js:220:30)"]}}},{"message":"Cannot return null for non-nullable field ProductItf.promiseNonNullErrorField","path":["allProducts",0]},{"message":"Cannot return null for non-nullable field ProductItf.promiseNonNullErrorField","path":["allProducts",1]},{"message":"Cannot return null for non-nullable field ProductItf.promiseNonNullErrorField","path":["allProducts",2]}]}]}
--graphql--

```

With a `products` array containing 3 items, the second chunk's `incremental` field contains 3 objects, each with `data: null` as expected as well as the correct `path`, but each object's `errors` array contains 6 errors: one per `product` returned (3) * 2 different ways the error is formatted. This results 18 errors total (the same 6 errors returned for each item).

CleanShot 2022-09-19 at 13 26 09@2x

There seems to be a second issue here as well which is that a single error is returned two different ways: `"INTERNAL_SERVER_ERROR"` with `stacktrace: "Error: Cannot return null for non-nullable field Product.promiseNonNullErrorField."` and `"Cannot return null for non-nullable field ProductItf.promiseNonNullErrorField"` (in our schema, `type Product implements ProductItf`).

**To Reproduce**
1. Checkout `https://github.com/apollographql/router-defer-e2e-tests/pull/8`
2. `cd CRA-demo && npm run router:up`
3. `npm i && npm start`
4. Visit http://localhost:3000/error-async-non-nullable-in-deferred-fragment and view response in network tab
5. `npm run router:down` to bring router down after testing

**Expected behavior**
Still determining the expected output per the spec.

**Desktop (please complete the following information):**

### PR DESCRIPTION
**Pull request by Geal**
Fix #1834 
Fix #1818

errors were wrongly copied on each incremental response. This will instead filter errors based on their path, and dispatch them to the relevant incremental response, while also removing the `_entities` field from the path

### CODE REVIEW

**Comment by bnjjj** [comment]
Would be happy to have a test for this

File: apollo-router/src/services/supergraph_service.rs [hunk 1]
```diff
@@ -507,4 +509,139 @@ mod tests {
...
+                    ]
+                },
+                "errors": [
+                    {
+                        "message": "error orga 1",
+                        "path": ["_entities", 0],
+                    },
+                    {
+                        "message": "error orga 3",
+                        "path": ["_entities", 2],
```
Comment (BrynCooke) [inline apollo-router/src/services/supergraph_service.rs:617]: Nit, orga 2?

File: apollo-router/src/query_planner/fetch.rs [hunk 2]
```diff
@@ -298,18 +298,43 @@ impl FetchNode {
...
-            .into_iter()
-            .map(|error| Error {
-                locations: error.locations,
-                path: error.path.map(|path| current_dir.join(path)),
-                message: error.message,
-                extensions: error.extensions,
-            })
-            .collect();
+        let mut inverted_paths: HashMap<usize, Vec<&Path>> = HashMap::new();
+        for (path, index) in paths.iter() {
```
Comment (BrynCooke) [inline apollo-router/src/query_planner/fetch.rs:302]: Nit, possibly this section could benefit from comments. The logic seems a bit tricky, and it took a while for me to figure out what it was doing.

**Review by BrynCooke — APPROVED** [review]

File: apollo-router/src/query_planner/fetch.rs (hunk 2 above)
Comment (Geal) [reply apollo-router/src/query_planner/fetch.rs:303]: right, I added more comments, the paths manipulations are confusing otherwise

**Pull request merged**